AtlanticOS: retain native proc/task allocation, RO metadata and lifetime operations.
These are licensed XNU function bodies, not BSD startup, userland or fake process providers.
Compile metadata with native BSD private headers and the accessor with Mach private headers.

diff --git a/osfmk/kern/task.h b/osfmk/kern/task.h
--- a/osfmk/kern/task.h
+++ b/osfmk/kern/task.h
@@ -326,7 +326,7 @@
 	uint32_t p_switch;            /* total processor switches */
 	uint32_t ps_switch;           /* total pset switches */
 
-#ifdef  MACH_BSD
+#if MACH_BSD || CONFIG_ATLANTIC_FIRMWARE
 	struct proc_ro *                bsd_info_ro;
 #endif
 	kcdata_descriptor_t             corpse_info;
diff --git a/osfmk/kern/atlantic_task_metadata.c b/osfmk/kern/atlantic_task_metadata.c
new file mode 100644
--- /dev/null
+++ b/osfmk/kern/atlantic_task_metadata.c
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2000-2020 Apple Inc. All rights reserved.
+ *
+ * @APPLE_OSREFERENCE_LICENSE_HEADER_START@
+ *
+ * This file contains Original Code and/or Modifications of Original Code
+ * as defined in and that are subject to the Apple Public Source License
+ * Version 2.0 (the 'License'). You may not use this file except in
+ * compliance with the License. The rights granted to you under the License
+ * may not be used to create, or enable the creation or redistribution of,
+ * unlawful or unlicensed copies of an Apple operating system, or to
+ * circumvent, violate, or enable the circumvention or violation of, any
+ * terms of an Apple operating system software license agreement.
+ *
+ * Please obtain a copy of the License at
+ * http://www.opensource.apple.com/apsl/ and read it before using this file.
+ *
+ * The Original Code and all software distributed under the License are
+ * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
+ * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
+ * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
+ * Please see the License for the specific language governing rights and
+ * limitations under the License.
+ *
+ * @APPLE_OSREFERENCE_LICENSE_HEADER_END@
+ */
+/* Copyright (c) 1995 NeXT Computer, Inc. All Rights Reserved */
+/*
+ * Copyright (c) 1982, 1986, 1989, 1991, 1993
+ *	The Regents of the University of California.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. All advertising materials mentioning features or use of this software
+ *    must display the following acknowledgement:
+ *	This product includes software developed by the University of
+ *	California, Berkeley and its contributors.
+ * 4. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ *	@(#)kern_proc.c	8.4 (Berkeley) 1/4/94
+ */
+/*
+ * NOTICE: This file was modified by SPARTA, Inc. in 2005 to introduce
+ * support for mandatory and extensible security protections.  This notice
+ * is included in support of clause 2.2 (b) of the Apple Public License,
+ * Version 2.0.
+ */
+/* HISTORY
+ *  04-Aug-97  Umesh Vaishampayan (umeshv@apple.com)
+ *	Added current_proc_EXTERNAL() function for the use of kernel
+ *      lodable modules.
+ *
+ *  05-Jun-95 Mac Gillon (mgillon) at NeXT
+ *	New version based on 3.3NS and 4.4
+ */
+
+
+
+/* AtlanticOS: pinned native proc/task metadata only; no BSD startup or userland.
+ * Function bodies below retain upstream lifetime, RO-zone and credential calls.
+ */
+#include <kern/task.h>
+#include <kern/zalloc.h>
+#include <sys/proc_internal.h>
+#include <sys/proc_require.h>
+#include <sys/proc_ro.h>
+#include <sys/kauth.h>
+#include <libkern/section_keywords.h>
+
+#if !CONFIG_ATLANTIC_FIRMWARE || MACH_BSD
+#error Atlantic task metadata subset requires the firmware profile without BSD userland
+#endif
+
+const size_t proc_alignment = _Alignof(struct proc);
+size_t proc_struct_size = sizeof(struct proc);
+size_t proc_and_task_size;
+ZONE_DECLARE_ID(ZONE_ID_PROC_TASK, struct proc);
+SECURITY_READ_ONLY_LATE(zone_t) proc_task_zone;
+SECURITY_READ_ONLY_LATE(proc_t) kernproc;
+ZONE_DEFINE_ID(ZONE_ID_PROC_RO, "proc_ro", struct proc_ro,
+    ZC_READONLY | ZC_ZFREE_CLEARMEM);
+
+task_t
+proc_get_task_raw(proc_t proc)
+{
+	return (task_t)((uintptr_t)proc + proc_struct_size);
+}
+
+proc_t
+task_get_proc_raw(task_t task)
+{
+	return (proc_t)((uintptr_t)task - proc_struct_size);
+}
+
+__attribute__((always_inline, visibility("hidden")))
+void
+proc_require(proc_t proc, proc_require_flags_t flags)
+{
+	if ((flags & PROC_REQUIRE_ALLOW_NULL) && proc == PROC_NULL) {
+		return;
+	}
+	zone_id_require(ZONE_ID_PROC_TASK, proc_and_task_size, proc);
+}
+
+static void
+proc_free(proc_t proc, proc_ro_t proc_ro)
+{
+	kauth_cred_t cred;
+
+	assert(proc_ro != NULL);
+
+	cred = smr_serialized_load(&proc_ro->p_ucred);
+	kauth_cred_set(&cred, NOCRED);
+
+	zfree_ro(ZONE_ID_PROC_RO, proc_ro);
+
+	zfree(proc_task_zone, proc);
+}
+
+void
+task_ref_hold_proc_task_struct(task_t task)
+{
+	proc_t proc_from_task = task_get_proc_raw(task);
+	os_atomic_or(&proc_from_task->p_refcount, P_REF_TASK_HOLD, relaxed);
+}
+
+void
+task_release_proc_task_struct(task_t task, proc_ro_t proc_ro)
+{
+	proc_t proc_from_task = task_get_proc_raw(task);
+	uint32_t old_ref = os_atomic_andnot_orig(&proc_from_task->p_refcount, P_REF_TASK_HOLD, relaxed);
+
+	if ((old_ref & P_REF_PROC_HOLD) == 0) {
+		proc_free(proc_from_task, proc_ro);
+	}
+}
+
+proc_ro_t
+proc_ro_alloc(proc_t p, proc_ro_data_t p_data, task_t t, task_ro_data_t t_data)
+{
+	proc_ro_t pr;
+	struct proc_ro pr_local = {};
+
+	pr = (proc_ro_t)zalloc_ro(ZONE_ID_PROC_RO, Z_WAITOK | Z_NOFAIL | Z_ZERO);
+
+	if (p != PROC_NULL) {
+		pr_local.pr_proc = p;
+		pr_local.proc_data = *p_data;
+	}
+
+	if (t != TASK_NULL) {
+		pr_local.pr_task = t;
+		pr_local.task_data = *t_data;
+	}
+
+	if ((p != PROC_NULL) || (t != TASK_NULL)) {
+		zalloc_ro_update_elem(ZONE_ID_PROC_RO, pr, &pr_local);
+	}
+
+	return pr;
+}
+
+proc_ro_t
+proc_ro_ref_task(proc_ro_t pr, task_t t, task_ro_data_t t_data)
+{
+	struct proc_ro pr_local;
+
+	if (pr->pr_task != TASK_NULL) {
+		panic("%s: proc_ro already has an owning task", __func__);
+	}
+
+	pr_local = *pr;
+	pr_local.pr_task = t;
+	pr_local.task_data = *t_data;
+
+	zalloc_ro_update_elem(ZONE_ID_PROC_RO, pr, &pr_local);
+
+	return pr;
+}
+
+void
+proc_ro_erase_task(proc_ro_t pr)
+{
+	zalloc_ro_update_field_atomic(ZONE_ID_PROC_RO,
+	    pr, pr_task, ZRO_ATOMIC_XCHG_LONG, TASK_NULL);
+}
+
+task_t
+proc_ro_task(proc_ro_t pr)
+{
+	return pr->pr_task;
+}
+
diff --git a/osfmk/kern/atlantic_task_process.c b/osfmk/kern/atlantic_task_process.c
new file mode 100644
--- /dev/null
+++ b/osfmk/kern/atlantic_task_process.c
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2000-2018 Apple Inc. All rights reserved.
+ *
+ * @APPLE_OSREFERENCE_LICENSE_HEADER_START@
+ *
+ * This file contains Original Code and/or Modifications of Original Code
+ * as defined in and that are subject to the Apple Public Source License
+ * Version 2.0 (the 'License'). You may not use this file except in
+ * compliance with the License. The rights granted to you under the License
+ * may not be used to create, or enable the creation or redistribution of,
+ * unlawful or unlicensed copies of an Apple operating system, or to
+ * circumvent, violate, or enable the circumvention or violation of, any
+ * terms of an Apple operating system software license agreement.
+ *
+ * Please obtain a copy of the License at
+ * http://www.opensource.apple.com/apsl/ and read it before using this file.
+ *
+ * The Original Code and all software distributed under the License are
+ * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
+ * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
+ * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
+ * Please see the License for the specific language governing rights and
+ * limitations under the License.
+ *
+ * @APPLE_OSREFERENCE_LICENSE_HEADER_END@
+ */
+
+#include <kern/task.h>
+#include <sys/proc_require.h>
+
+void  *
+get_bsdtask_info(task_t t)
+{
+	void *proc_from_task = task_get_proc_raw(t);
+	proc_require(proc_from_task, PROC_REQUIRE_ALLOW_NULL | PROC_REQUIRE_ALLOW_ALL);
+	return task_has_proc(t) ? proc_from_task : NULL;
+}
+
